server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -174,34 +177,35 @@ private synchronized void closeWithTragicEvent(final Exception ex) {
      * @return the location the bytes were written to
      * @throws IOException if writing to the translog resulted in an I/O exception
      */
-    public synchronized Translog.Location add(final BytesReference data, final long seqNo) throws IOException {
-        ensureOpen();
-        final long offset = totalOffset;
-        try {
-            data.writeTo(outputStream);
-        } catch (final Exception ex) {
-            closeWithTragicEvent(ex);
-            throw ex;
-        }
-        totalOffset += data.length();
+    public Translog.Location add(final ReleasableBytesReference data, final long seqNo) throws IOException {
+        final Translog.Location location;
+        final long bytesBufferedAfterAdd;
+        synchronized (this) {
+            ensureOpen();
+            final long offset = totalOffset;
+            totalOffset += data.length();
+            bufferedBytes += data.length();
+            bufferedOps.add(new Operation(seqNo, data.retain()));
 
-        if (minSeqNo == SequenceNumbers.NO_OPS_PERFORMED) {
-            assert operationCounter == 0;
-        }
-        if (maxSeqNo == SequenceNumbers.NO_OPS_PERFORMED) {
-            assert operationCounter == 0;
-        }
+            assert minSeqNo != SequenceNumbers.NO_OPS_PERFORMED || operationCounter == 0;
+            assert maxSeqNo != SequenceNumbers.NO_OPS_PERFORMED || operationCounter == 0;
 
-        minSeqNo = SequenceNumbers.min(minSeqNo, seqNo);
-        maxSeqNo = SequenceNumbers.max(maxSeqNo, seqNo);
+            minSeqNo = SequenceNumbers.min(minSeqNo, seqNo);
+            maxSeqNo = SequenceNumbers.max(maxSeqNo, seqNo);
 
-        nonFsyncedSequenceNumbers.add(seqNo);
+            operationCounter++;
 
-        operationCounter++;
+            assert assertNoSeqNumberConflict(seqNo, data);
 
-        assert assertNoSeqNumberConflict(seqNo, data);
+            location = new Translog.Location(generation, offset, data.length());
+            bytesBufferedAfterAdd = bufferedBytes;
+        }
+
+        if (bytesBufferedAfterAdd >= FORCE_WRITE_THRESHOLD) {
+            writeBufferedOps(Long.MAX_VALUE);
```

Comment:
I had intended to move this to a conditional lock in a follow-up, but since you raised the point here I made that change in this PR. The only intention of this work is to flush the translog if it gets too large.